### ISSUE TITLE
Missing Normalization Step after adding Gaussian Noise

### ISSUE THREAD

**Issue by PHANTOM0122** [liw95/LightLoc#5]
Dear author.

Hi, thank you for your great work.
I think the normalization step after adding gaussian noise to global embedding as mentioned in the paper is missing in current train.py.
Is the paper is wrong or just missing in your implementation?

If the paper is right, please check my pull request.

Thanks.

Best,
Heejoon.

**Comment by liw95** [comment]
Hi Heejoon,

Thank you for your careful reading and helpful comment!
I tested the implementation with the additional normalization step on the QEOxford dataset. The resulting localization errors for the four trajectories are 1.08 m / 1.24°, 1.08 m / 1.16°, 0.96 m / 1.16°, and 1.03 m / 1.20°, which are slightly worse than the original results.

Therefore, please take the current implementation (without the extra normalization) as the correct one.

Best regards,
Wen Li

**Comment by liw95** [comment]
Specifically, the implementation uses the backbone and classification head weights from the Model Zoo, and trains the classification head with an additional normalization step.

**Comment by PHANTOM0122** [comment]
Thanks for your detailed reply.
I also had the same degradation when adding the code line into trian.py.

B.t.w., regarding training regression head with backbone and classification head from the Model Zoo,
I had a similar results with the paper when using the classification head from the Model zoo,
but suffers from slight reimplementaion error when training classification head also from scratch.

Below is my results of training the classification head and the regression head.
<img width="570" height="149" alt="Image" src="https://github.com/user-attachments/assets/3a35308e-b98c-4d53-b052-1176e10dc313" />

I think this might be from the version of package (e.g. scikit-learn for K-means clustering) version. 
Can I get detailed package version used in the environment and QEOxford_cls_pose_stats.txt / NCLT_cls_pose_stats.txt?

Thanks.
Heejoon.

**Comment by liw95** [comment]
Thanks for sharing your results.
The scikit-learn version we used is 1.5.1.
From your results, it seems you have successfully reproduced the results of the paper.
You may try training the classifier multiple times, since the classification head can show slight variation due to random initialization.

**Comment by PHANTOM0122** [comment]
Thanks again for your kind reply!

It seems that sckit-learn==1.5.1 is available at  python >= 3.10,
which is not compatible with python=3.8 mentioned in the repository's setup instruction.

Can I get the conda enviroment.yml file used in the paper experiment?
It will be helpful to my research, and I'll be really appreciate about it!
(You can also send the file to my email: wilko97@hanyang.ac.kr )

Best,
Heejoon.

**Comment by liw95** [comment]
Hi, my apologies. The version of scikit-learn is 1.3.2. You can download the environment.yml file from the attachment.

[environment.yaml](https://github.com/user-attachments/files/23140188/environment.yaml)

**Comment by PHANTOM0122** [comment]
Thank you for your helpful commitment.
It works well for me. 

Lastly, can you share the code for training the backbone??

**Comment by liw95** [comment]
Hi, sorry for the late reply. Thank you for your interest. Unfortunately, the training code for the backbone is not planned to be open-sourced at the moment.
